### ISSUE TITLE
No error when casting an array of tuples

### ISSUE THREAD

**Issue by ivan7237d** [microsoft/TypeScript#41341]
**TypeScript Version:** 4.0.3

**Search Terms:**
cast tuple

**Code**

```ts
[[], [1]] as [number][];
```

**Expected behavior:**
Error would be produced.

**Actual behavior:**
No error. If this is the expected behavior, what is the workaround in the context of functional-style code that uses expressions and avoids statements like `const a: [number][] = [[], [1]];`? The best I can think of is defining 
```ts
const identity = <T>(value: T) => value;
```
and then using it like this: 
```ts
identity<[number][]>([[], [1]]);
```
which does produce an error but clutters up the generated JS code.

**Playground Link:** 
https://www.typescriptlang.org/play?#code/MYewdgzgLgBAhjAvDA2igugGlQRne+CVMAVwFsAjAUwCd0MBuIA

**Related Issues:**

**Comment by RyanCavanaugh** [comment]
Casts are allowed to downcast, so this is a legal assertion.

> what is the workaround in the context of functional-style code

Generally if you're writing in this style, there's a place to put generic type arguments for whatever function you're passing this to. I don't have more specific guidance without the context of a larger sample.

See also #7481

**Comment by ivan7237d** [comment]
Thanks @RyanCavanaugh. What's at the root of this issue is that for `[[], [1]]` the compiler infers type `number[]` whereas as I was expecting it to do two things that it doesn't do right now: (1) infer the type of `[[], [1]]` as `[[], [number]]` and (2) not allow to cast that type to `[number][]` because `[number]` isn't a subtype of `[] | [number]`.

You've answered my question, but just by way of feedback, the reason why generic type arguments in existing functions don't work for me and I have to use identity<T> is that I'm using a function `applyPipe(x, a, b)` which is an equivalent of `b(a(x))` or `x |> a |> b`. That function is defined with a bunch of overloads for different numbers of arguments (like `pipe` in RxJS), and has a generic type for each argument, so I can't use applyPipe<MyType>. BTW if in the future TS starts supporting the pipeline operator, there won't even be a function to specify generic type for, so +1 for the issue you referenced.
